CI scripts using GHA, fixes to code and tests and refactoring

cgr71ii opened the pull request:
Changes:

- 3 different CI workflows have been implemented: min tests, min tests without python dependencies and intensive tests (this last workflow will be executed not after each commit but every weekly or manually). The workflow for the intensive tests is not for sure that it works, since it will be possible to test it [once it's reached at least once the master branch](https://github.com/orgs/community/discussions/24657).
- Installation was broken since there were incompatible dependencies:
  - `ftfy` with different versions in 2 `requirements.txt` files.
  - `requetsts` had incompatibilities with other requirements.
  - `linguacrawl` has been disabled (check out [this commit](https://github.com/bitextor/bitextor/commit/313e621af87b89188b076c29f29a4f982b3fbf12) for further explanation).
- Submodules have been moved to `third_party` following the same schema that Bitextor does (it makes the repository cleaner and it's something other repositories do like [pytorch](https://github.com/pytorch/pytorch/tree/master/third_party)).
- Tests have been fixed since they were not working or were adapted to bitextor.

Observations:

- Documentation is being updated.
- Different TODO's have been added to the Snakefile since there are parts of the code which I think are not doing what is expected.
- `run-deferred-tests.sh` has been partially adapted to Monotextor, but the reconstructor doesn't have support for mono version as far as I know.
- Running the tests in local, only `run-tests-min.sh` success, and the reason seems to be related to issue #2. This leads to the CI workflows to fail as well but the one that executed `run-tests-min.sh` (you can check it out, for instance, in [this commit](https://github.com/bitextor/monotextor/commit/9c96853058501e48566288dac1e24a9faf7ba9dd)).

cgr71ii committed:
Move submodules to third_party

Minor documentation update

cgr71ii committed:
Bump versions and fix installation

cgr71ii committed:
Fixes

cgr71ii committed:
Remove __pycache__

cgr71ii committed:
Add GHA script

cgr71ii committed:
Fix installation documentation and GHA script

Monocleaner needs hunspell

cgr71ii committed:
GHA script typo

cgr71ii committed:
Update tests and new GHA script

cgr71ii committed:
New GHA script

cgr71ii committed:
Fix tests

cgr71ii committed:
Minor tests fix

cgr71ii committed:
GHA script without python dependencies

cgr71ii commented:
This PR should be reviewed and accepted/rejected once #2 had been solved.

cgr71ii commented:
Related to one of my TODO's which refers to the cleaning and post-processing

```
# TODO further rules should use shard and batches, and if we want to reduce only use batches, a rule like split_segalign from bitextor might be used
```

there is a problem: the transient directory is not being used.

The transient directory should be used for files like Monocleaner (is [the way Bitextor uses this directory](https://github.com/bitextor/bitextor/blob/5cfec4d7f242dcda72e3b7fdc049a6660dbad300/bitextor/Snakefile#L1352)), but since it seems that the post-processing is not being correctly applied, at least if we take Bitextor as reference, this directory is not being used properly. Should an issue be created?

I think this might be related to another TODO that I've put related to the rule `aggregate_split`:
```
# TODO this is not the expected behaviour of rule "aggregate_*", at least if we compare with Bitextor
```

This rule is being used for aggregate all the content in one file, what is not what is applied in Bitextor. When this is done, we lose the possibility of parallelization for the last steps, which is something that Bitextor does.

cgr71ii committed:
Update documentation

cgr71ii committed:
Update documentation

ZJaume commented on tests/common.sh:
```diff
@@ -2,46 +2,131 @@
     fi
 }
 
 download_monocleaner_model()
```
Monocleaner has `monocleaner-download` command in case you want to simplify this function.

cgr71ii replied:
I didn't know, and it's a good idea (thank you for the tip!), but I think that for this case it'd be better to keep this way in order to avoid the tests to fail if, for some reason, monocleaner is not installed. I know it will fail anyway, but the specific tests which use monocleaner, not all of them.

ZJaume commented on monotextor/Snakefile:
```diff
@@ -880,41 +879,34 @@ rule monocleaner:
-    filter_input = rules.monocleaner.input.monofixer
-
-
+# TODO is correct to apply deduplication in the rule filter? If so, there is no difference between rule raw and sents when PARAGRAPHS="" (i.e. the output files of the rules raw an sents will have the exact same content)
```
Yes, it is correct to apply dedup here. Regarding the difference of the rules, they seem to be exact the same shell code. But sents produces a deduped and filtered output, and raw produces unfiltered and undeduped. Inputs are not the same.

cgr71ii replied:
Ah, ok! When I was making tests I noticed that produced the same files when paragraphs were disabled. I guess that was coincidence that there were not duplicate sentences.

ZJaume reviewed (approved):
For me, all the changes seem ok besides a couple of comments.

cgr71ii commented:
There are a few points that I think they need to be discussed (`TODO`s in the `Snakefile`):

- Variable `DISABLE_MONOCLEANER_HARDRULES` is intended to be used? Is not being used as far as I've checked.
  - There are other variables which are not used and I don't know if they've been forgotten or they're intended to be used (e.g. `WORDTOKS`, `MORPHTOKS`).
  - `SENTTOKS` and `CUSTOMNBPS` (I might've missed some others) are used, but are not in `args.py`, so they can't be provided through configuration. Are they intended to be used or these options should be removed?
- Rule `aggregate_split`: "this is not the expected behaviour of rule 'aggregate_*', at least if we compare with Bitextor". Indeed, the description of the rule is "Helper rule to implement until=split config", but the current behavior is, if I'm not wrong, write a file with the format `URL<tab>paragraph`.
- When `aggregate_split` rule is reached, the sharding is totally removed, so we lose parallelism. In Bitextor is kept until final files, so we can still get profit from parallelism if we keep the batches and resplit the batches like is done in Bitextor in rule `split_segalign`.

ZJaume commented:
`DISABLE_HARDRULES` was used until we disabled them completely for all cases. I think it can be removed. `WORDTOKS` and `MORPHTOKS`  are not used because there is no tokenization in any step.

`SENTTOKS` and `CUSTOMNBPS` are there in case we add support for another sentence splitter than loomchild. `CUSTOMNBPS` would be used if we add Moses, I think. We should keep these, at least for now, and add them to `args` when we add a new splitter in the future.

`aggregate_split` just merges all the shards and batches because they aren't needed for the rest of the pipeline. Parallelism is not lost because the rest of the rules only need to run gnu parallel inside and that's enough. Maybe the documentation of the rule is misleading and should be updated, but the rule is fine, I think.

cgr71ii committed:
Minor fixes and update TODOs

cgr71ii committed:
Use transient dir and minor fix

04_03.filtered typo when UNTIL == 'filter'

cgr71ii commented:
Thank you, @ZJaume! I've updated the code according to your answer.

Regarding the use of `parallel`, it's true that can handle parallelism, but the way I say and that's being used in Bitextor, once the files reach Bifixer, or Monofixer in this case, files are merged and split again. Doing this, we have batches, which can be parallelized at job level, and after that, `parallel` is used, but in files which are not as big as they could be if this steps was not being used. Since this is not used in Monotextor and only `parallel` is used instead of job parallelism + `parallel`, the level of parallelism is lower. Of course, this is not essential, so I think that it may let it be as it's now.

Once you can review the new changes and #2 is resolved (I'd like to test that all tests pass), I think the PR will be ready.

cgr71ii committed:
Update submodules

cgr71ii committed:
Fix sentence splitter and monocleaner for empty files

Fix https://github.com/bitextor/monotextor/issues/2

cgr71ii committed:
Update CREDITS

cgr71ii commented:
I think the PR is ready for merge once reviewed!

ZJaume commented on docs/OUTPUT.md:
```diff
@@ -0,0 +1,35 @@
+        * `monofixer_score` rates quality of duplicate or near-duplicate text
+    5. `monocleaner_lang_id monocleaner_score` - Monocleaner classifer output
+        * `monocleaner_lang_id` is the lang which Monocleaner detects
+        * `monocleaner_score` is the quality metric of Monocleaner for the text
```
Could it be changed to "fluency score" instead of "quality metric"?

For the lang id maybe it's worth mentioning [FastSpell](https://github.com/mbanon/fastspell)

cgr71ii replied:
Done!

cgr71ii committed:
Update OUTPUT.md
